File: Examples/RudderSampleAppObjC/RudderSampleAppObjC/EncryptedDatabaseProvider.m
//
//  EncryptedDatabaseProvider.m
//  RudderDatabaseEncryption
//
//  Created by Pallab Maiti on 14/09/23.
//

#import "EncryptedDatabaseProvider.h"
#import "sqlite3.h"

@interface RSEncryptedDatabase : NSObject <RSDatabase>

@end

@implementation RSEncryptedDatabase {
    sqlite3 *db;
}

- (int)open_v2:(const char *)filename flags:(int)flags zVfs:(const char *)zVfs {
    return sqlite3_open_v2(filename, &db, flags, zVfs);
}


- (int)exec:(const char *)zSql xCallback:(callback)xCallback pArg:(void *)pArg pzErrMsg:(char * _Nullable *)pzErrMsg {
    return sqlite3_exec(db, zSql, xCallback, pArg, pzErrMsg);
}

- (int)close {
    return sqlite3_close(db);
}

- (int)step:(void *)pStmt {
    return sqlite3_step(pStmt);
}

- (int)finalize:(void *)pStmt {
    return sqlite3_finalize(pStmt);
}

- (int)prepare_v2:(const char *)zSql nBytes:(int)nBytes ppStmt:(void **)ppStmt pzTail:(const char **)pzTail {
    return sqlite3_prepare_v2(db, zSql, nBytes, (sqlite3_stmt **)(ppStmt), pzTail);
}

- (int)column_int:(void *)pStmt i:(int)i {
    return sqlite3_column_int(pStmt, i);
}

- (const unsigned char *)column_text:(void *)pStmt i:(int)i {
    return sqlite3_column_text(pStmt, i);
}

- (int)key:(const void *)pKey nKey:(int)nKey {
    return sqlite3_key(db, pKey, nKey);
}

- (int)last_insert_rowid {
    int64_t lastRowId = sqlite3_last_insert_rowid(db);
    return (int)lastRowId;
}

@end

@implementation EncryptedDatabaseProvider

- (id<RSDatabase>)getDatabase {
    return [RSEncryptedDatabase new];
}

@end

File: Examples/RudderSampleAppSwift/RudderSampleAppSwift/EncryptedDatabaseProvider.swift
//
//  EncryptedDatabaseProvider.swift
//  RudderSampleAppSwift
//
//  Created by Pallab Maiti on 14/09/23.
//  Copyright © 2023 RudderStack. All rights reserved.
//

import Foundation
import Rudder

class EncryptedDatabase: RSDatabase {
    
    private var db: OpaquePointer?
    
    func open_v2(_ filename: UnsafePointer<CChar>?, flags: Int32, zVfs: UnsafePointer<CChar>?) -> Int32 {
        return sqlite3_open_v2(filename, &db, flags, zVfs)
    }
    
    func exec(_ zSql: UnsafePointer<CChar>?, xCallback: callback?, pArg: UnsafeMutableRawPointer?, pzErrMsg: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?) -> Int32 {
        return sqlite3_exec(db, zSql, xCallback, pArg, pzErrMsg)
    }
        
    func prepare_v2(_ zSql: UnsafePointer<CChar>, nBytes: Int32, ppStmt: UnsafeMutablePointer<UnsafeMutableRawPointer?>?, pzTail: UnsafeMutablePointer<UnsafePointer<CChar>?>?) -> Int32 {
        return sqlite3_prepare_v2(db, zSql, nBytes, UnsafeMutablePointer(OpaquePointer(ppStmt)), pzTail)
    }
    
    func close() -> Int32 {
        return sqlite3_close(db)
    }
    
    func step(_ pStmt: UnsafeMutableRawPointer?) -> Int32 {
        return sqlite3_step(OpaquePointer(pStmt))
    }
    
    func finalize(_ pStmt: UnsafeMutableRawPointer?) -> Int32 {
        return sqlite3_finalize(OpaquePointer(pStmt))
    }
    
    func column_int(_ pStmt: UnsafeMutableRawPointer?, i: Int32) -> Int32 {
        return sqlite3_column_int(OpaquePointer(pStmt), i)
    }
    
    func column_text(_ pStmt: UnsafeMutableRawPointer?, i: Int32) -> UnsafePointer<UInt8> {
        return sqlite3_column_text(OpaquePointer(pStmt), i)
    }
    
    func key(_ pKey: UnsafeRawPointer?, nKey: Int32) -> Int32 {
        return sqlite3_key(db, pKey, nKey)
    }
}

class EncryptedDatabaseProvider: RSDatabaseProvider {
    func getDatabase() -> RSDatabase {
        return EncryptedDatabase()
    }
}
